libs/server-sdk/src/evaluation/evaluator.hpp

```diff
@@ -6,33 +6,37 @@
 #include <launchdarkly/logging/logger.hpp>
 #include <launchdarkly/value.hpp>
 
-#include "bucketing.hpp"
-#include "detail/evaluation_stack.hpp"
 #include "evaluation_error.hpp"
+#include "evaluation_stack.hpp"
 
 #include "../data_interfaces/store/istore.hpp"
 #include "../events/event_scope.hpp"
 
-#include <tl/expected.hpp>
-
 namespace launchdarkly::server_side::evaluation {
 
 class Evaluator {
    public:
+    /**
+     * Constructs a new Evaluator. Since the Evaluator may be used by multiple
+     * threads in parallel, the given logger and IStore must be thread safe.
+     * @param logger A logger for recording errors or warnings.
+     * @param source The flag/segment source.
+     */
     Evaluator(Logger& logger, data_interfaces::IStore const& source);
 
     /**
      * Evaluates a flag for a given context.
-     * Warning: not thread safe.
      *
      * @param flag The flag to evaluate.
      * @param context The context to evaluate the flag against.
+     * @param stack The evaluation stack used for detecting circular references.
      * @param event_scope The event scope used for recording prerequisite
      * events.
      */
     [[nodiscard]] EvaluationDetail<Value> Evaluate(
         data_model::Flag const& flag,
         Context const& context,
+        EvaluationStack& stack,
```

Comment:
Is there a reason the public API needs to take a stack versus making one. Or are we re-using the public version internal to the evaluator as well?

Reply:
We also may want to consider a more generic `EvaluationState` that contains the evaluation stack. 

Eventually you have big segment status and big segment membership that persist the duration of an evaluation.

Reply:
Yes, we are using this in AllFlags. So we setup the stack object once, then run all the evaluations on it. 

I think I did that because it felt bad to be constructing the stack over and over for each flag - but I didn't have any data, so I'd be fine pushing it back into the Evaluator. 

It'd be safer that way too, if the stack somehow got un-balanced by a failed evaluation.